Support IDENTITY() function in SELECT-INTO

deepakshi-mittal opened the pull request:
Changes for Identity in Select Into statement
Added hooks for identity
Added fields in intoclause to pass information about identity column

Engine Pr: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/166

Task : BABEL-539

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deepakshi-mittal committed:
Support IDENTITY() function in SELECT-INTO

Changes for Identity in Select Into statement
Added hooks for identity
Added fields in intoclause to pass information about identity column

Task : BABEL-539

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Added test to all upgrade files

Added syntax error if query tries to calls internal function

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

macwoody commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3391,6 +3391,9 @@ CREATE OR REPLACE FUNCTION sys.host_id()
 
+CREATE OR REPLACE FUNCTION sys.identity_into(IN typename INT, IN seed INT, IN increment INT)
+RETURNS int AS 'babelfishpg_tsql' LANGUAGE C STABLE;
+
```
Here is a dilemma you are facing. 
If you don't grant permission to public, non-dbo users would not be able to call this function. Thus could not use your feature. 
If you grant, everyone would be able to call, which would move a sequence forward bypassing security check.

forestkeeper replied:
could be 
1. grant to public , and then 
2. add a block in parser to block customer from calling this func directly

deepakshi-mittal replied:
Done 
grant to public , and then
add a block in parser to block customer from calling this func directly

macwoody commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -230,6 +230,36 @@ TsqlFunctionConvert(TypeName *typename, Node *arg, Node *style, bool try, int lo
+			case INT4OID:
+			case INT8OID:
+			case NUMERICOID:				
+				args = list_make3((Node *)makeIntConst((int)type_oid, -1), seed, increment);
```
why -1?

deepakshi-mittal replied:
its for location, I will change it

deepakshi-mittal committed:
Added missing function for failing upgrade test

Task: BABEL-539

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
+			type_string = "pg_catalog.numeric";
+			break;
+		default:
+			type_string = "decimal";
```
why do we have decimal as default? Shouldn't it be an error here?

forestkeeper replied:
+1
Should throw error for other types, and add a case for decimal

macwoody commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1846,6 +1847,17 @@ class tsqlBuilder : public tsqlCommonMutator
 
+	void enterFunction_call(TSqlParser::Function_callContext *ctx) override
+	{
+		std::string func_name = ::getFullText(ctx);
```
I think you don't have to do this way. 
There is more elegant way of getting function name. 
You could find helper functions to do so.

deepakshi-mittal replied:
Done used helper functions

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
+	return type_string;
+}
+
+static List * transformSelectIntoStmt(CreateTableAsStmt *stmt, const char *queryString){
```
Please try drop dependent sequence explicitly and see if the dropping sequence could be blocked. 
And also check if the sequence is gone after dropping the table.

deepakshi-mittal replied:
Yes when we try to drop sequence explicitly it errors out saying that can't drop a dependent sequence.
Yes sequence is gone after we drop the table.

deepakshi-mittal committed:
Added missing function for failing Validate test

Task: BABEL-539

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
+		{	
+			TargetEntry *tle = (TargetEntry *) lfirst(elements);
+			FuncExpr *funcexpr;
+			if (tle->expr && IsA(tle->expr, FuncExpr) && strcasecmp(get_func_name(((FuncExpr *) (tle->expr))->funcid), "identity_into") ==0 ){
```
we should write 
```
if ()
{
}
```

not 
```
if(){
}
```
please make { not the same line with if ()

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
+						case 3:
+							con = (Const *) fargNode;
+							seqoptions = lappend(seqoptions,  makeDefElem("increment", (Node *)makeInteger(con->constvalue), -1));
+							if ((int) con->constvalue > 0){
```
Should update the if else format to 
```
if()
{
}
else
{
}
```

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
+	ObjectAddress secondaryObject = InvalidObjectAddress;
+	List *stmts;
+	stmts = transformSelectIntoStmt((CreateTableAsStmt *) parsetree, queryString);
+		while (stmts != NIL)
```
the tab interval is not right from this line

deepakshi-mittal committed:
Code review comments

Corrected formatting
Added case for decimal
used inbuilt function for getting procedure name
Added syntax error if identity_into function is called directly

Task: BABEL-539

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Revert changes for failing test

Not required because no added test cases for callin these functions

Task: BABEL-539

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper reviewed (approved):
lgtm
